GitHub CI

ggwpez opened the pull request:
Waiting for the `txCache` fix.

sebastianst commented on operator/init_test.go:
```diff
@@ -3,6 +3,6 @@ package operator
 
 func init() {
-	logrus.SetLevel(logrus.InfoLevel)
+	logrus.SetLevel(logrus.TraceLevel)
```
I wouldn't change that in master.

ggwpez replied:
I just saw that the `tee/prototype/init_test.go` is set to `TraceLevel`, we could also change it there.

sebastianst commented on operator/config.go:
```diff
@@ -33,3 +35,24 @@ func LoadConfig(path string) (*Config, error) {
+// envGanacheCmdVar is the environment variable to set the command
+// for spawning a `ganache-cli`. It is set "ERDSTALL_GANACHE_CMD".
+// See `ganacheCommand()`.
+const envGanacheCmdVar = "ERDSTALL_GANACHE_CMD"
```
The `Var` is superfluous.
```suggestion
const envGanacheCmd = "ERDSTALL_GANACHE_CMD"
```

sebastianst commented on operator/operator_test.go:
```diff
@@ -122,18 +122,23 @@ func initEnvironment(t *testing.T) *environment {
 	var cmd *exec.Cmd
 	runGanache := true
 	if runGanache {
```
Please boyscout-fix this. Either just remove the bool and execute the indented block directly or pass `runGanache bool` as a boolean flag to `initEnvironment`. Not sure if there's a use case where we don't want to start it though. @matthiasgeihs what was the intention here?

ggwpez replied:
That would actually have been much easier than to inject the spawn command via an `env` variable...  
I will remove it for now since it works and is easier to use. When we want it in the future we can still add it.   

Let's keep this open until @matthiasgeihs replies.

sebastianst replied:
I think it's still best to pass the ganach cmd as an env variable because that's easiest to configure in CI configs (like in your github one)

matthiasgeihs replied:
this was for local testing reasons when i wanted to run ganache separately. i'm fine with removing that option.

ggwpez replied:
Ok good

sebastianst commented on operator/operator_test.go:
```diff
@@ -178,7 +183,9 @@ func (e *environment) WaitPhase() {
 func (e *environment) Shutdown() {
 	if e.cmd != nil {
-		e.cmd.Process.Kill()
+		if err := e.cmd.Process.Kill(); err != nil {
+			log.Warn("Could not kill process:", err)
+		}
```
you could use `e.t` to assert that this doesn't fail.

ggwpez replied:
This will be executed in the test cleanup, so i dont think we should/need to let it fail here.

sebastianst commented on operator/test/user.go:
```diff
@@ -63,6 +63,7 @@ func CreateUser(
 ) *User {
 	ethClient, err := eth.CreateEthereumClient(ethURL, wallet, account)
 	if err != nil {
+		panic(err)
 		t.Fatal("creating ethereum wallet and client:", err)
```
Now the `Fatal` call is not reached. `Fatal` would crash the application anyways, why the panic?

sebastianst commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,37 @@
+      run: |
+        go get -v -t -d ./...
+        docker pull trufflesuite/ganache-cli
+
```
We should also add `golangci-lint` here.

sebastianst replied:
With a configuration similar to our go-perun one.

ggwpez replied:
I used the default config now, should be good.

sebastianst commented on operator/operator_test.go:
```diff
@@ -138,7 +138,7 @@ func initEnvironment(t *testing.T) *environment {
 			panic(err)
 		}
-		time.Sleep(5 * time.Second)
+		time.Sleep(10 * time.Second)
```
Instead of a blind sleep, we could run a loop here that tries to connect to the ganache in, say .2 sec steps, for 10 sec. Would speed up the test in most cases. Or maybe instead of a loop, just one request with a 10 sec timeout, not sure what works best.

You could just do a generic query like `curl -X POST --data '{"jsonrpc":"2.0","method":"eth_blockNumber","params":[],"id":99999}'`, see https://eth.wiki/json-rpc/API#eth_blocknumber or `eth_protocolVersion`, but not using curl but Go's internal http requests.

ggwpez replied:
Good idea, i will look into it once i fixed all the other things.

ggwpez replied:
I added a retry loop for the dial.

sebastianst commented on client/client.go:
```diff
@@ -369,24 +364,6 @@ func (c *Client) CmdDeposit(status chan *CmdStatus, args ...string) {
-	return
-}
-
-func errorReason(ctx context.Context, b *pethchannel.ContractBackend, tx *types.Transaction, blockNum *big.Int, acc accounts.Account) (string, error) {
```
do you already use our `ContractBackend.WaitMined` function, which tries to extract revert reasons?

ggwpez replied:
This function was unused, therefore i deleted it. The client+operator currently use `abi/bind.WaitMined` but should probably switch to `ContractBackend.ConfirmTransaction`.

sebastianst replied:
and we should probably just rename it to `WaitMined` in go-perun :joy:

sebastianst replied:
-> #54

sebastianst commented on operator/operator_test.go:
```diff
@@ -183,9 +183,7 @@ func (e *environment) WaitPhase() {
-		if err := e.cmd.Process.Kill(); err != nil {
-			log.Warn("Could not kill process:", err)
-		}
+		e.cmd.Process.Kill() // nolint: errcheck
```
What's the linter's problem here? Wouldn't it be better to log the error?

ggwpez replied:
It collided with a change from Matthias, i will opt for the log.

sebastianst commented:
You can delete the Travis pipeline.

sebastianst commented on cmd/client/main.go:
```diff
@@ -5,6 +5,7 @@ package main
 import (
 	perunchannel "perun.network/go-perun/backend/ethereum/channel"
 	perunhd "perun.network/go-perun/backend/ethereum/wallet/hd"
+	"perun.network/go-perun/log"
```
we use logrus, not perun log

sebastianst commented on operator/operator_test.go:
```diff
@@ -119,21 +119,21 @@ func initEnvironment(t *testing.T) *environment {
+		fmt.Sprintf("--mnemonic=\"%s\"", cfg.Mnemonic),
+		fmt.Sprintf("--blockTime=%d", blockTime),
+	)
+	cmd := exec.Command(
+		prog,
+		args...,
+	)
```
merge to one line

sebastianst commented on operator/operator_test.go:
```diff
@@ -112,7 +112,7 @@ type environment struct {
 }
 
-const blockTime = 1 // block mining interval in seconds
+const blockTime = 2 // block mining interval in seconds
```
Can we change the test so that we subscribe to new heads to synchronize with the ganache's blocks? This way we can reuse the same test for testnets.

And I think that it makes sense to have the ganache use a block time of 1 sec so that tests run faster.

ggwpez replied:
Yea we have a `SubscribeToBlocks` function in the ethereum client which should work.

ggwpez replied:
Changed it back to 1 sec.

sebastianst commented on eth/client.go:
```diff
@@ -105,12 +106,21 @@ func (cl *Client) NewTransactor(ctx context.Context) (*bind.TransactOpts, error)
+		if err != nil {
+			select {
+			case <-ctx.Done():
+				return nil, fmt.Errorf("dialing ethereum node: %v", err)
```
```suggestion
				return nil, fmt.Errorf("dialing ethereum node: %w", err)
```

sebastianst commented on operator/operator_test.go:
```diff
@@ -164,11 +164,17 @@ func initEnvironment(t *testing.T) *environment {
 func (e *environment) WaitPhase() {
-	waitBlock := func() {
-		time.Sleep(blockTime * time.Second)
+	sub, err := e.operator.ethClient.SubscribeToBlocks()
```
This blocks subscription is unnecessarily complex because it retrieves _all_ receipts of the whole block. Please just make it work with a simple header subscription, the only information you need is the block number.

ggwpez replied:
I tried it before with a header subscription, but it did not work. I can try it again if you think it is worth it.

ggwpez replied:
Ok it seems to work now, i probably did not have the `-1` there while trying it before.

sebastianst commented on operator/operator_test.go:
```diff
@@ -164,11 +164,17 @@ func initEnvironment(t *testing.T) *environment {
+	defer sub.Unsubscribe()
+
+	for block := range sub.Blocks() {
+		// - 1 to make sure that we are in the new phase.
+		if e.enclaveParameters.IsLastPhaseBlock(block.NumberU64() - 1) {
```
How is `WaitPhase` used? It's confusing to see this `-1` because now you're waiting until the first block of a new phase has been mined, not when a phase starts.

ggwpez replied:
It did not work without the `-1`. Maybe @matthiasgeihs can tell us what it was actually meant for.

sebastianst replied:
I implemented `IsLastPhaseBlock` and it does
```go
// IsLastPhaseBlock tells whether this block is the last block of a phase.
func (p Parameters) IsLastPhaseBlock(blockNum uint64) bool {
	return ((blockNum - p.InitBlock) % p.PhaseDuration) == p.PhaseDuration-1
}
```
so from my understanding, it is `true` for the last block of any phase. I.e., if the last block is _mined_ then the phase is done.

ggwpez committed:
:rotating_light: Fix all linter warnings.

Signed-off-by: Oliver Tale-Yazdi <oliver@perun.network>

ggwpez committed:
:construction_worker: Add GitHub CI.

The CI will mainly be used to execute the on-chain operator test
against a local ganache.

Signed-off-by: Oliver Tale-Yazdi <oliver@perun.network>

ggwpez committed:
:sparkles: [op] Retry failed eth-node connections.

In order to get a connection to the eth-node as soon as it comes
online, we use a loop which retries a dial call until it succeeds.
This is more reasonable than a banal wait.

Signed-off-by: Oliver Tale-Yazdi <oliver@perun.network>

ggwpez committed:
:sparkles: [op] Use ENV variable to pass ganache-cli command.

ERDSTALL_GANACHE_CMD can now be used to set override the default
ganache-cli spawn command.

Signed-off-by: Oliver Tale-Yazdi <oliver@perun.network>

ggwpez committed:
:white_check_mark: [op] Use block wait instead of time.Sleep.

Signed-off-by: Oliver Tale-Yazdi <oliver@perun.network>

sebastianst commented on operator/operator_test.go:
```diff
@@ -164,11 +166,25 @@ func initEnvironment(t *testing.T) *environment {
 func (e *environment) WaitPhase() {
-	waitBlock := func() {
-		time.Sleep(blockTime * time.Second)
+	ctx, cancel := context.WithTimeout(context.Background(), blockTime*time.Duration(e.cfg.PhaseDuration+1))
```
PhaseDuration is in number of blocks _not_ a time. This here would be a few milli seconds, so always be a done context the moment you use it.

ggwpez committed:
:white_check_mark: [op] Stabilize OP test.

Signed-off-by: Oliver Tale-Yazdi <oliver@perun.network>

sebastianst committed:
tee/proto: Add logging to Enclave.Shutdown

sebastianst committed:
tee/proto: Use logrus logger in enclave tests

Before, `t.Log` was used, which doesn't show up in CI logs.

sebastianst committed:
eth: Add SimSetup.Balance to retrieve user balances

sebastianst committed:
tee/proto: Fix exiter logging

Don't take Hash() of common.Address when logging.

sebastianst committed:
tee/proto: Improve enclave test to check final balance proofs

This also helps synchronizing the enclave routine and block subscription
so that the subscription is not closed before all mined blocks have been
processed.

sebastianst commented on operator/operator_test.go:
```diff
@@ -178,7 +182,36 @@ func (e *environment) WaitPhase() {
+		case head := <-heads:
+			n--
+			if n == 0 {
+				log.Warn("Phase ended in block #", head.Number.Uint64())
```
I wouldn't put a `Warn` log here. Also, `WaitBlock` doesn't know why it is called, so it cannot say "Phase ended".

sebastianst commented on operator/operator_test.go:
```diff
@@ -178,7 +182,36 @@ func (e *environment) WaitPhase() {
 		case head := <-heads:
-			if e.enclaveParameters.IsLastPhaseBlock(head.Number.Uint64() - 1) {
+			if e.enclaveParameters.IsLastPhaseBlock(head.Number.Uint64()) {
+				log.Warn("Phase ended in block #", head.Number.Uint64())
```
debug log

ggwpez replied:
The stupid `Warn` calls are only my way of Debug logging since i do not need to change the loglevel to see them :laughing:  
Will surely remove them.

sebastianst commented on operator/operator.go:
```diff
@@ -95,6 +95,7 @@ func Setup(cfg *Config) *Operator {
 		PhaseDuration:    cfg.PhaseDuration,
 		ResponseDuration: cfg.ResponseDuration,
 		PowDepth:         cfg.PowDepth,
+		InitBlock:        1,
```
You don't need that, `DeployContracts` should set the `InitBlock`. Also, you cannot know the correct init block number here.

ggwpez replied:
I will remove it.
